### DISCUSSION TITLE
Cross compiling for ESP32/ESP8266

### DISCUSSION THREAD

**Discussion by capsulecorplab** [nasa/fprime#1612] [General]
I have a few use cases where I'd like to be able to run fprime on an ESP32 or ESP8266 device, s.a., a D1 mini. Any idea if this is possible?

**Comment by SterlingPeet** [comment]
I also have an interest in this, I have a high confidence that it could work.  Caveat Emptor, I have flown missions with F Prime on a ATmega128, so I know small platforms can be done and they require a fair amount of platform work to shoehorn it in there.

**Reply by capsulecorplab** [reply]
@SterlingPeet what was your process for designing a build system for the atmega? I've been working on a containerized environment for doing fprime development on https://github.com/capsulecorplab/kasm-fprime-workspace that I'm updating to include build tools for platforms s.a., AVR and ESP devices. If of interest, maybe we can collaborate on this effort

**Reply by SterlingPeet** [reply]
I haven’t gotten into containerization yet, partly because all my different projects are rooted in systems that predate that popularity (at least in the space circles I run in).  I usually develop direct on OS X / macOS, but many developers I support use a virtualbox image that I developed to run on VMware, using Linux as the VM OS.  That has worked quite well in-house.

**Reply by capsulecorplab** [reply]
@SterlingPeet Feel free to ping me if you'd like a demo of the virtual image I've been building for fprime development. A core feature about the environment I'm currently using is that the system level tools/utilities in the environment are immutable and configured/built from an Ansible playbook.yaml file. It also runs locally on docker, so distributed team members can essentially run the same development environment across platforms.

**Reply by YeltsinMartin** [reply]
@SterlingPeet Could you please share the repo that ports fprime to Arduino? I am looking to run fprime on Arduino Mega 2560.

**Reply by SterlingPeet** [reply]
Well, admittedly this fell off the radar.

@capsulecorplab yes I would like to see that, preferably as a 1 page style write up  with pointers to docs.

@LeStarch we should probably get back to wrapping up the ``arduino-cli`` project so we can make that available in the community repos.

@YeltsinMartin do you have an SRAM expansion board for your Mega 2560?  You can run maybe one or two components on the 8k internal SRAM but you can't run any of the commanding or telemetry components without an expansion to 64k.

**Reply by LeStarch** [reply]
@SterlingPeet I am going to be looking at producing a version 0.1 with Arduino CLI (hopefully) over the weekend and moving it to fprime-community's fprime-arduino repo.  The we can move forward from there?

**Reply by SterlingPeet** [reply]
Yes, @LeStarch thats a good plan.

**Reply by YeltsinMartin** [reply]
@SterlingPeet I don't have one, I have some spare ESP8266's would that work ?

**Reply by SterlingPeet** [reply]
@YeltsinMartin I haven't been able to get full system under the 32k limit that ESP8266 incurs, although I have not tried compiling for that platform explicitly.  Same problem for the Arduino SAMD21, which I have attempted to compile and it is too big for the platform.  I think that the ESP32s might be possible, but I also have not tried directly.

**Reply by YeltsinMartin** [reply]
@SterlingPeet thanks for the info!

**Reply by capsulecorplab** [reply]
@SterlingPeet Let me know if the minimal README in this repo will suffice https://github.com/capsulecorplab/fprime-workspace-image
